Document Elastic Stack configuration policies (#6215)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/stack-config-policy.asciidoc

:page_id: stack-config-policy
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{page_id}.html[View this document on the Elastic website]
****
endif::[]
[id="{p}-{page_id}"]
= Elastic Stack configuration policies

NOTE: This requires a valid Enterprise license or Enterprise trial license. Check <<{p}-licensing,the license documentation>> for more details about managing licenses.

> Starting from ECK `2.6.0` and Elasticsearch `8.6.0`, Elastic Stack configuration policies let you configure in a declarative way, for multiple clusters, the following settings:

Reviewer: ```suggestion
Starting from ECK `2.6.0` and Elasticsearch `8.6.0`, Elastic Stack configuration policies allow you to configure the following settings:
```

declarative is implied imo, and multiple clusters is mentioned after the settings list again

- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/put-pipeline-api.html[Ingest pipelines]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-put-lifecycle.html[Index Lifecycle Policies]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-template.html[Index templates]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-component-template.html[Components templates]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-put-role-mapping.html[Role mappings]

A policy can be applied to one or more Elasticsearch clusters in any namespace managed by the ECK operator.

Configuration policy settings applied by the ECK operator are immutable through the Elasticsearch REST API.

It is currently not allowed to configure an Elasticsearch cluster with more than one policy.

[float]
[id="{p}-{page_id}-definition"]
== Define Elastic Stack configuration policies

Elastic Stack configuration policies can be defined in a `StackConfigPolicy` resource. Each `StackConfigPolicy` must have the following fields:

* `name` is a unique name used to identify the policy.

> * `spec.elasticsearch` describes the settings to configure and at least one setting must be defined. Each of the following fields is a named entity map:

Reviewer: "named entity map" is not self-explanatory. Also I don't think it is true for the `clusterSettings`?

Author: True. What do you thing of:

> Each of the following fields except `clusterSettings` is an associative array where the keys are arbitrary names and the values are the definitions:

>   ** `clusterSettings` are the settings commonly found in the elasticsearch.yaml file.

Reviewer: ```suggestion
  ** `clusterSettings` are the settings that go into the elasticsearch.yml file.
```
or something like that. The thing that threw me off was "commonly found"

  ** `snapshotRepositories` are snapshot repositories for defining an off-cluster storage location for your snapshots.
  ** `snapshotLifecyclePolicies` are snapshot lifecycle policies, to automatically take snapshots and control how long they are retained.
  ** `securityRoleMappings` are role mappings, to define which roles are assigned to each user by identifying them through rules.
  ** `ingestPipelines` are ingest pipelines, to perform common transformations on your data before indexing.
  ** `indexLifecyclePolicies` are index lifecycle policies, to automatically manage the index lifecycle.
  ** `indexTemplates.componentTemplates` are component templates that are building blocks for constructing index templates that specify index mappings, settings, and aliases.
  ** `indexTemplates.composableIndexTemplates` are index templates to define settings, mappings, and aliases that can be applied automatically to new indices.

The following fields are optional:

> * `namespace` is the namespace of the `StackConfigPolicy` resource and used to identify the Elasticsearch clusters to which this policy applies. If it equals to the operator namespace, all namespaces managed by the operator are used, otherwise only the policy namespace is used.

Reviewer: ```suggestion
* `namespace` is the namespace of the `StackConfigPolicy` resource and used to identify the Elasticsearch clusters to which this policy applies. If it equals to the operator namespace, the policy applies to all namespaces managed by the operator, otherwise the policy only applies to the namespace of the policy.
```

* `resourceSelector` is a link:https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/[label selector] to identify the Elasticsearch clusters to which this policy applies in combination with the namespace(s). No `resourceSelector` means all Elasticsearch clusters in the namespace(s).
* `secureSettings` is a list of Secrets containing Secure Settings to inject into the keystore(s) of the Elasticsearch cluster(s) to which this policy applies, similar to the<<{p}-custom-images,Elasticsearch Secure Settings>>.

> Secure settings are required to configure Cloud snapshot repositories (azure, gcs, s3) if you are not using Workload Identity (e.g.: https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity).

Reviewer: AWS IAM roles for service accounts is a similar feature on AWS that allows users to avoid having to specify secure settings. https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-snapshots.html#k8s-iam-service-accounts

Author: I thought "Workload Identity" might be the generic term for this feature but not really and I forgot we already documented it so well. I will make the link between the two docs.

```suggestion
Secure settings are required to configure Cloud snapshot repositories (azure, gcs, s3) if you are not using Cloud-provider specific means to leverage Kubernetes service accounts
(<<{p}-gke-workload-identiy,GKE Workload Identity>> or <<{p}-iam-service-accounts,AWS IAM roles for service accounts>>, for example).

[source,yaml,subs="attributes,+macros"]
----

Example of applying a policy that configures snapshot repository, SLM Policies, and cluster settings:

apiVersion: stackconfigpolicy.k8s.elastic.co/v1alpha1
kind: StackConfigPolicy
metadata:
  name: test-stack-config-policy
  # namespace: elastic-system or test-namespace
spec:
  resourceSelector:
    matchLabels:
      env: my-label
  secureSettings:
  - secretName: "my-secure-settings"
  elasticsearch:
    clusterSettings:
      indices.recovery.max_bytes_per_sec: "100mb"
    snapshotRepositories:
      test-repo:
        type: gcs
        settings:
          bucket: my-bucket
    snapshotLifecyclePolicies:
      test-slm:
        schedule: "0 1 2 3 4 ?"
        name: "<production-snap-{now/d}>"
        repository: test-repo
        config:
          indices: ["*"]
          ignore_unavailable: true
          include_global_state: false
        retention:
          expire_after: "7d"
          min_count: 1
          max_count: 20
----

> Another example of configuring role mapping, ingest pipeline, ILM and index templates:

Reviewer: ```suggestion
Another example of configuring role mappings, ingest pipelines, ILM and index templates:
```

                max_num_segments: 1
            min_age: 10d
    indexTemplates:
      componentTemplates:
        test-component-template:
          template:
            mappings:
              properties:
                '@timestamp':
                  type: date
        test-runtime-component-template-test:
          template:
            mappings:
              runtime:
                day_of_week:
                  type: keyword
      composableIndexTemplates:
        test-template:
          composed_of:
          - test-component-template
          - test-runtime-component-template-test
          index_patterns:
          - test*
          - bar*
          priority: 500
          template:
            aliases:
              mydata: {}
            mappings:
              _source:
                enabled: true
              properties:
                created_at:
                  format: EEE MMM dd HH:mm:ss Z yyyy
                  type: date
                host_name:
                  type: keyword
            settings:
              number_of_shards: 1
          version: 1
----

[float]
[id="{p}-{page_id}-monitoring"]
== Monitor Elastic Stack configuration policies

> In addition to the logs generated by the operator, a config policy status is maintained in the `StackConfigPolicy` resource. This status summarizes the health of the configuration policy through a phase and the number of resources for which the policy could be applied.

Reviewer: ```suggestion
In addition to the logs generated by the operator, a config policy status is maintained in the `StackConfigPolicy` resource. This status gives information in which phase the policy is ( "Applying", "Ready", "Error") and it indicates the number of resources for which the policy could be applied.
```

Starting from ECK `2.6.0` and Elasticsearch `8.6.0`, Elastic Stack configuration policies let you configure in a declarative way, for multiple clusters, the following settings:

- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/settings.html[Cluster Settings]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/put-snapshot-repo-api.html[Snapshot Repositories]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/slm-api-put-policy.html[Snapshot Lifecycle Policies]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/put-pipeline-api.html[Ingest pipelines]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-put-lifecycle.html[Index Lifecycle Policies]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-template.html[Index templates]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-component-template.html[Components templates]
- link:https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-put-role-mapping.html[Role mappings]

> A policy can be applied to all or a subset of Elasticsearch clusters in a given namespace or all namespaces managed by the operator.

Reviewer 2: ```suggestion
A policy can be applied to one or more Elasticsearch clusters in any namespace managed by the ECK operator.
```

> Settings described in a policy applied by the ECK operator are immutable through the Elasticsearch REST API.

Reviewer 2: ```suggestion
Configuration policy settings applied by the ECK operator are immutable through the Elasticsearch REST API.
```

> It is not allowed to configure an Elasticsearch cluster with more than one policy currently.

Reviewer 2: ```suggestion
It is currently not allowed to configure an Elasticsearch cluster with more than one policy.
```

[float]
[id="{p}-{page_id}-definition"]
== Define Elastic Stack configuration policies

> Elastic Stack configuration policies can be defined in an `StackConfigPolicy` resource. Each `StackConfigPolicy` must have the following fields:

Reviewer 2: ```suggestion
Elastic Stack configuration policies can be defined in a `StackConfigPolicy` resource. Each `StackConfigPolicy` must have the following fields:
```

* `name` is a unique name used to identify the policy.

> * `spec.elasticsearch` describes the settings to configure, at least one setting must be defined. Each following field is a named entity map:

Reviewer 2: ```suggestion
* `spec.elasticsearch` describes the settings to configure and at least one setting must be defined. Each of the following fields is a named entity map:
```

>   ** `clusterSettings` are the settings commonly found in elasticsearch.yaml settings.

Reviewer 2: ```suggestion
  ** `clusterSettings` are the settings commonly found in the elasticsearch.yaml file.
```

>   ** `snapshotRepositories` are snapshot repositories, to define off-cluster storage location for your snapshots.

Reviewer 2: ```suggestion
  ** `snapshotRepositories` are snapshot repositories for defining an off-cluster storage location for your snapshots.
```

  ** `snapshotLifecyclePolicies` are snapshot lifecycle policies, to automatically take snapshots and control how long they are retained.
  ** `securityRoleMappings` are role mappings, to define which roles are assigned to each user by identifying them through rules.
  ** `ingestPipelines` are ingest pipelines, to perform common transformations on your data before indexing.
  ** `indexLifecyclePolicies` are index lifecycle policies, to automatically manage the index lifecycle.
  ** `indexTemplates.componentTemplates` are component templates that are building blocks for constructing index templates that specify index mappings, settings, and aliases.
  ** `indexTemplates.composableIndexTemplates` are index templates to define settings, mappings, and aliases that can be applied automatically to new indices.

The following fields are optional:

> * `namespace` is the namespace of the `StackConfigPolicy` resource, it is used to identify the Elasticsearch clusters to which this policy applies. If it equals to the operator namespace, all namespaces managed by the operator are used, otherwise only the policy namespace is used.

Reviewer 2: ```suggestion
* `namespace` is the namespace of the `StackConfigPolicy` resource and used to identify the Elasticsearch clusters to which this policy applies. If it equals to the operator namespace, all namespaces managed by the operator are used, otherwise only the policy namespace is used.
```

* `resourceSelector` is a link:https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/[label selector] to identify the Elasticsearch clusters to which this policy applies in combination with the namespace(s). No `resourceSelector` means all Elasticsearch clusters in the namespace(s).

> * `secureSettings` is a list of Secrets containing Secure Settings to inject into the keystore of the Elasticsearch to which this policy applies, similar to the<<{p}-custom-images,Elasticsearch Secure Settings>>.

Reviewer 2: ```suggestion
* `secureSettings` is a list of Secrets containing Secure Settings to inject into the keystore(s) of the Elasticsearch cluster(s) to which this policy applies, similar to the<<{p}-custom-images,Elasticsearch Secure Settings>>.
```

            min_age: 30d
          warm:
            actions:
              forcemerge:
                max_num_segments: 1
            min_age: 10d
    indexTemplates:
      componentTemplates:
        test-component-template:
          template:
            mappings:
              properties:
                '@timestamp':
                  type: date
        test-runtime-component-template-test:
          template:
            mappings:
              runtime:
                day_of_week:
                  type: keyword
      composableIndexTemplates:
        test-template:
          composed_of:
          - test-component-template
          - test-runtime-component-template-test
          index_patterns:
          - test*
          - bar*
          priority: 500
          template:
            aliases:
              mydata: {}
            mappings:
              _source:
                enabled: true
              properties:
                created_at:
                  format: EEE MMM dd HH:mm:ss Z yyyy
                  type: date
                host_name:
                  type: keyword
            settings:
              number_of_shards: 1
          version: 1
----

> 

Reviewer 2: I think there's an extra newline here.

[float]
[id="{p}-{page_id}-monitoring"]
== Monitor Elastic Stack configuration policies

In addition to the logs generated by the operator, a config policy status is maintained in the `StackConfigPolicy` resource. This status summarizes the health of the configuration policy through a phase and the number of resources for which the policy could be applied.

[source,sh]
----
kubectl get stackconfigpolicy
----

[source,sh]
----
NAME                           READY   PHASE   AGE
test-stack-config-policy       1/1     Ready   1m42s

> test-err-stack-config-policy   0/1     Error   1m42s

Reviewer 2: Should these be combined into the same `[source,sh]`?

Author: I don't have a strong opinion, but in most pages of the doc, we keep the two separate (e.g.: [quickstart](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-deploy-elasticsearch.html#k8s-elasticsearch-monitor-cluster-health), [autoscaling](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-autoscaling.html#k8s-monitoring), [commons-problems](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-common-problems.html)).

Reviewer 2: No problem then @thbkrkr .  Just from reading it, I thought they should be combined as one is the command, and the other is the output of the same command, but I don't have a strong opinion either, tbh.

----

When not all resources are ready, you can get more information about the reason by reading the full status:

[source,sh]
----

> kubectl get -n b scp test-2-stack-config-policy -o jsonpath="{.status}" | jq .

Reviewer 2: Should this name `test-2-stackc-config-policy` be referencing the one in error above?  `test-err-stack-config-policy`?
